File: code/1729. Transforming Operators RxSwift-3-02/RxSwift-3-02/20-transforming-operators-starter/RxSwiftPlayground/RxSwiftPlayground.playground/Sources/SupportCode.swift
import Foundation
import RxSwift

public func example(of description: String, action: () -> Void) {
  print("\n--- Example of:", description, "---")
  action()
}

public let episodeI = "Episode I - The Phantom Menace"
public let episodeII = "Episode II - Attack of the Clones"
public let episodeIII = "Episode III - Revenge of the Sith"
public let episodeIV = "Episode IV - A New Hope"
public let episodeV = "Episode V - The Empire Strikes Back"
public let episodeVI = "Episode VI - Return Of The Jedi"
public let episodeVII = "Episode VII - The Force Awakens"
public let episodeVIII = "Episode VIII - The Last Jedi"
public let episodeIX = "Episode IX"

public let episodes = [episodeI, episodeII, episodeIII, episodeIV, episodeV, episodeVI, episodeVII, episodeVIII, episodeIX]

public extension String {
  
  /// https://stackoverflow.com/a/36949832/616764
  func romanNumeralIntValue() throws -> Int  {
    if range(of: "^(?=[MDCLXVI])M*(C[MD]|D?C{0,3})(X[CL]|L?X{0,3})(I[XV]|V?I{0,3})$", options: .regularExpression) == nil  {
      throw NSError(domain: "NotValidRomanNumber", code: -1, userInfo: nil)
    }
    
    var result = 0
    var maxValue = 0
    
    uppercased().reversed().forEach {
      let value: Int
      switch $0 {
      case "M":
        value = 1000
      case "D":
        value = 500
      case "C":
        value = 100
      case "L":
        value = 50
      case "X":
        value = 10
      case "V":
        value = 5
      case "I":
        value = 1
      default:
        value = 0
      }
      
      maxValue = max(value, maxValue)
      result += value == maxValue ? value : -value
    }
    
    return result
  }
}

public struct Jedi {
  
  public var rank: BehaviorSubject<JediRank>
  
  public init(rank: BehaviorSubject<JediRank>) {
    self.rank = rank
  }
}

public enum JediRank: String {
  
  case youngling = "Youngling"
  case padawan = "Padawan"
  case jediKnight = "Jedi Knight"
  case jediMaster = "Jedi Master"
  case jediGrandMaster = "Jedi Grand Master"
}

File: code/2161. Transforming Operators 3-002-transforming-operators/3-002-transforming-operators/recording/end/src/main/kotlin/SupportCode.kt
import io.reactivex.subjects.BehaviorSubject

fun exampleOf(description: String, action: () -> Unit) {
  println("\n--- Example of: $description ---")
  action()
}

const val episodeI = "Episode I - The Phantom Menace"
const val episodeII = "Episode II - Attack of the Clones"
const val episodeIII = "Episode III - Revenge of the Sith"
const val episodeIV = "Episode IV - A New Hope"
const val episodeV = "Episode V - The Empire Strikes Back"
const val episodeVI = "Episode VI - Return Of The Jedi"
const val episodeVII = "Episode VII - The Force Awakens"
const val episodeVIII = "Episode VIII - The Last Jedi"
const val episodeIX = "Episode IX - Episode IX"

val episodes = listOf(episodeI, episodeII, episodeIII, episodeIV, episodeV, episodeVI, episodeVII, episodeVIII, episodeIX)

fun String.romanNumeralIntValue(): Int {

  // https://gist.github.com/eadred/a902ec625f2ecb0a98841e78e16438d6

  val lookup: Map<Char, Int> = mapOf(
      Pair('I', 1),
      Pair('V', 5),
      Pair('X', 10),
      Pair('L', 50),
      Pair('C', 100),
      Pair('D', 500),
      Pair('M', 1000)
  )

  fun accumulate(currentValue: Int?, lastValueAndTotal: Pair<Int, Int>): Pair<Int, Int> {
    if (currentValue == null) return lastValueAndTotal
    val mult = if (lastValueAndTotal.first > currentValue) -1 else 1
    return Pair(currentValue, lastValueAndTotal.second + (mult * currentValue))
  }

  return this.map { lookup[it] }
      .foldRight(Pair(0, 0), ::accumulate)
      .second
}

data class Jedi(val rank: BehaviorSubject<JediRank>)

enum class JediRank(val value: String) {
  Youngling("Youngling"),
  Padawan("Padawan"),
  JediKnight("Jedi Knight"),
  JediMaster("Jedi Master"),
  JediGrandMaster("Jedi Grand Master")
}